### PR TITLE
Release 2024-04-11(ish) cherry-pick #7348

### PR DESCRIPTION
**Pull request by sharnoff**
## Release 2024-04-11(ish), #7348

Cherry-picked a single PR - #7348. See there, or neondatabase/cloud#12047 for more.

ref https://neondb.slack.com/archives/C04DGM6SMTM/p1712776981582679

### CODE REVIEW

File: compute_tools/src/config.rs [hunk 1]
```diff
@@ -92,6 +92,27 @@ pub fn write_postgres_conf(
...
+        // dynamic_shared_memory_type = 'mmap'.
+        //
+        // This is (maybe?) temporary - for more, see https://github.com/neondatabase/cloud/issues/12047.
+        let overcommit_memory_contents = std::fs::read_to_string("/proc/sys/vm/overcommit_memory")
+            // ignore any errors - they may be expected to occur under certain situations (e.g. when
+            // not running in Linux).
+            .unwrap_or_else(|_| String::new());
+        if overcommit_memory_contents.trim() == "2" {
+            let opt = GenericOption {
+                name: "dynamic_shared_memory_type".to_owned(),
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply compute_tools/src/config.rs:107]: Fair, yeah- previously, control plane specified this iff swap is enabled, but this PR is the compute_ctl side of moving the logic here. (See neondatabase/cloud#12097 for the cplane side).

I'm not sure how our user-provided settings works, or if we allow them to set this.

[the remark this answers is not in the document]
Comment (sharnoff) [reply compute_tools/src/config.rs:107]: Checked internally, it should not be possible for a user to set `dynamic_shared_memory_type/`.

[the remark this answers is not in the document]
Comment (lubennikovaav) [reply compute_tools/src/config.rs:107]: >I'm not sure how our user-provided settings works.

Users can provide some pgSettings via API: https://api-docs.neon.tech/reference/createprojectendpoint

but only some of them are allowed: see [allowed_settings](https://github.com/neondatabase/cloud/blob/main/goapp/internal/dto/postgres/pg_settings.go#L49)

I've tested `dynamic_shared_memory_type`

```
{
  "code": "",
  "message": "setting 'dynamic_shared_memory_type' cannot be changed"
}
```

[the remark this answers is not in the document]
Comment (lubennikovaav) [reply compute_tools/src/config.rs:107]: >Just to clarify — what would the result be if we also have dynamic_shared_memory_type=posix in the config?

you mean postgresql.conf content, right?
" If the file contains multiple entries for the same parameter, all but the last one are ignored."

[the remark this answers is not in the document]
Comment (sharnoff) [reply compute_tools/src/config.rs:107]: Ah awesome, thanks @lubennikovaav !

> by "config" here you mean postgresql.conf content, right?

Yeah

**Pull request merged**